### PR TITLE
Assignment 6 - Timothy N & Roberto G

### PR DESCRIPTION
**Pull request by timmytaters**
@timmytaters @robertogarcia11

### CODE REVIEW

**Comment by colestewart12** [comment]
I can review this PR!

File: Calculate.Tests/CalculatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    [TestMethod]
+    public void TryCalculate_NotEnoughArgs_Fails()
+    {
+        Calculate calculator = new();
+        int? answer;
+        Assert.IsFalse(calculator.TryCalculate(" 3 *", out answer));
+        Assert.IsNull(answer);
+    }
+}
```
Comment (colestewart12) [inline Calculate.Tests/CalculatorTests.cs:72]: Tests look good so far but may want to write tests to cover MathematicalOperations

**Review by colestewart12 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔
Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.✔
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall very well done! Code is clean and properly implemented, could use some more tests on the Calculate Class but other than that no big suggestions.

File: Calculate/Calculate.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    catch (FormatException)
+                    {
+                        throw new FormatException($"Unable to parse operands {components[0]} or {components[2]}");
+                    }
+                }
+            }
+        }
+
+        result = null;
+        return false;
```
Comment (omar5o9) [inline Calculate/Calculate.cs:71]: Good job on this method

File: Calculate.Tests/ProgramTests.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    [TestMethod]
+    public void Program_ChangeOutput_Succeeds()
+    {
+        Program program = new()
+        {
+            WriteLine = TestOutput
+        };
+        program.WriteLine("Yes");
+        Assert.AreEqual<string>("Yes", Output);
+    }
```
Comment (omar5o9) [inline Calculate.Tests/ProgramTests.cs:40]: Good job on these tests

**Review by omar5o9 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console.✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ❌✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Review by MannyPoptart — COMMENTED** [review]
Instructions

    Create a Console project called "Calculate.". ✔
    Define a Program Class
        Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
        Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
        Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
    Define a Calculator class ✔
        Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
        Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
            is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
                Uses char for the key corresponding to the operators +, -, *, and /. ✔
                Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
        Implement a TryCalculate method following "TryParse" pattern ✔
            Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
            If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
            Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
            Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
    Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
    Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Fundamentals

    Place all shared project properties into a Directory.Build.Props file. ✔
    Place all shared project items into a Directory.Build.targets file. ❌ (This is missing)
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ❌
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔ (Nice work on your program tests, though might need some more tests in CalculatorTests)
    Choose simplicity over complexity ✔

File: Calculate/Calculate.cs (hunk 2 above)
Comment (robertogarcia11) [reply Calculate/Calculate.cs:71]: omg thanks 😎

File: Calculate.Tests/ProgramTests.cs (hunk 3 above)
Comment (robertogarcia11) [reply Calculate.Tests/ProgramTests.cs:40]: thank you!

File: Calculate/Calculate.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ❌✔
+If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ❌✔
+Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ❌✔
+Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ❌✔*/
+
+namespace Calculate;
+
+public class Calculate
+{
+    public IReadOnlyDictionary<char, Func<int, int, int>> MathematicalOperations { get; } =
```
[the remark this answers is not in the document]
Comment (robertogarcia11) [reply Calculate/Calculate.cs:17]: thank you!

File: Calculate/Program.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace Calculate;
+
+public class Program
+{
+
+    // These properties are guaranteed to be non-null because of the 'init' setter,
+    // which mandates that they are initialized either during object instantiation or within a constructor.
+    public Action<string> WriteLine { get; init; } = Console.WriteLine;
+    public Func<string?> ReadLine { get; init; } = Console.ReadLine;
```
[the remark this answers is not in the document]
Comment (robertogarcia11) [reply Calculate/Program.cs:14]: 😎😁

**Comment by timmytaters** [comment]
> Place all shared project items into a Directory.Build.targets file. ❌ (This is missing)
> and enabled .NET analyzers for both projects ❌

The Directory.Build.targets file is missing because to my knowledge, there are no shared targets, and .NET analyzers are enabled in the Directory.Build.props

File: Calculate/Program.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            program.WriteLine("Please enter the problem (num operator num): ");
+            if(program.ReadLine == null)
+            {
+                input = "";
+            }
+            else
+            {
+#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type, okay since null checked above
+                input = program.ReadLine();
+#pragma warning restore CS8600
```
[the remark this answers is not in the document]
Comment (timmytaters) [reply Calculate/Program.cs:37]: Doing input = program.ReadLine() ?? ""; worked, thank you!

File: Calculate.Tests/CalculatorTests.cs (hunk 1 above)
Comment (timmytaters) [reply Calculate.Tests/CalculatorTests.cs:72]: According to my coverage extension, they are covered

File: Calculate.Tests/ProgramTests.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        Program program = new()
+        {
+            ReadLine = TestInput
+        };
+        Calculate calculator = new();
+        string? input = program.ReadLine();
+
+#pragma warning disable CS8604 // Possible null reference argument, if input null will throw error anyway, shouldn't be for premise
+        calculator.TryCalculate(input, out answer);
+#pragma warning restore CS8604
```
Comment (BenjaminMichaelis) [inline Calculate.Tests/ProgramTests.cs:28]: Disabling the warning is not the right solution, the compiler knows what it is talking about

File: Calculate/Program.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    public static void Main(string[] args)
+    {
+        Program program = new();
+        Calculate calculator = new();
+        string input;
+        int? answer;
+
+        do
+#pragma warning disable CS8604 // Possible null reference argument, input never null since either empty string or readline (when readline isnt null)
```
Comment (BenjaminMichaelis) [inline Calculate/Program.cs:26]: Disabling the warning is not the right solution, the compiler knows what it is talking about

File: Calculate/Calculate.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    public static int Addition(int a, int b) => a + b;
+    public static int Subtraction(int a, int b) => a - b;
+    public static int Multiplication(int a, int b) => a * b;
+
+    public static int Division(int dividend, int divisor)
+    {
+        if (divisor == 0)
+        {
+            throw new ArgumentException("Division by zero is undefined", nameof(divisor));
+        }
```
Comment (BenjaminMichaelis) [inline Calculate/Calculate.cs:35]: Divide by zero by default throws a `System.DivideByZeroException`

**Pull request closed without merging**